### ISSUE TITLE
VB needs faster CInt

### ISSUE THREAD

**Issue by ljw1004** [dotnet/roslyn#6603]
This topic was discussed in [VB LDM 2014-03-12](http://roslyn.codeplex.com/discussions/543475).

**Background**

``` cs
double d = 10.5;
int x = (int)d;
```

``` vb
Dim x = CInt(d)
```

The VB version is slower than C#. Indeed, any time we call `CInt`, the compiler always codegens a call to `Math.Round`. Even `CInt(Math.Round(x))` gets compiled into `Math.Round(Math.Round(x))`.

**Request**
Please allow us to write fast integer casts in VB.

**Resolution**

_Part 1_: Any time the compiler generates an implicit call to `Math.Round` (i.e. as part of `CInt`, `CLong`, `CULong`, `CUint`), and if the argument of that implicit call is one of a list of math functions known to return a whole number, then omit the outer call to `Math.Round`. The list of known math functions is: `Round(d)`, `Round(d,MidpointRounding)`, `Truncate(d)`, `Floor(d)`, `Ceiling(d)`, `VisualBasic.Fix(d)`, `VisualBasic.Int(d)`. 

_Part 2_: If the user writes `CInt(Math.Truncate(d))` then we will emit `conv.i4` or `conv.ovf.i4` dependent on the current project-level "check-overflow" settings.

_TODO_: before we making this fix, you must verify that `conv.i4` has exact same semantics as `CInt(Math.Truncate)`, including negative values, superlarge doubles and exceptions, NaN, Infinity). Also do likewise for `conv.i1`, `conv.i2`, `conv.i4`, `conv.i8` and for their unsigned versions, and for both checked and unchecked versions.

**Discussion**
This is a compiler optimization, pure and simple. It adds no new syntax or concepts or library functions. We identify targeted scenarios and generate optimal IL in those cases where semantics would not be affected. This seemed the cleanest approach. Specifically, it seemed better than adding any of the following syntaxes:

``` vb
DirectCast(d, Integer)
(Integer)d
VB.FastCast(d)
d As Integer
```

Will it be better to special-case just one of these combinations? e.g. `conv.i4` on `CInt(Truncate(double))`? Or is it better to just do all of them? Default answer: "yes do all of them", and only scale back if the implementor comes back and says it's too hard. 

There are other possible concerns, e.g. what happens if it turns out that the semantics of `conv.i4` are actually different from ALL of the Math.Round-like functions? What if it turns out that it's really hard to implement every case? What if expression-trees throw a wrench in the works? Well, when someone sets about implementing the feature, then we can circle back to check with the VB Language Design Meeting.

**Comment by JohnHamby** [comment]
Optimizing CInt(Math.Truncate(d)) is certainly better than the current situation, but it's neither discoverable nor obvious in reading that CInt(Math.Truncate(d)) would provide significantly better performance than CInt(d).

Another possible alternative is to add an optional operand to CInt, CType, and maybe various other C*** operators to control the semantics of the conversion. For controlling conversions from floating point to integral types, some possible values of the additional operand are Round (which gives the default semantics), Truncate or Chop (giving truncation semantics which can be optimized on current hardware), and maybe Fastest (giving rounding or truncation, whichever is fastest on the target architecture). Because this is new language, the semantic concerns raised in the last paragraph above do not arise.

IntelliSense would show the optional operand, making the mechanism discoverable. It's certainly easy to read:

    CInt(d, Round)
    CInt(d, Truncate)
    CInt(d, Fastest)
    CType(d, Integer, Truncate)

**Comment by DualBrain** [comment]
The only comments that I can add without fully understanding what means to be added to the 2.0 milestone and seeing the 2.0 (Preview) text under the milestone section is that:

I feel that focusing on the most occurring scenario (both personal experience and anecdotal discussions with others), optimizing the CInt(Fix(value)) pattern would make the most sense.  It pulls back the scope and would provide the biggest bang for the buck (AFAIK).

To reply to the other comment (and to agree further with the original post); I don't believe that this is a scenario that dictates any language changes.  It should not be something that "has to be discovered".  It should just work.  For those interested in knowing more about the man behind the mirror, those people will have no problem discovering what "tricks" there may be to milk out every bit of performance they can.  See the Select statement for more details. ;-)

I also agree that it needs to use conv.i4 or conv.ovf.i4 (if that is the correct counterpart - not an IL expert) based on project settings.

(I'm a relative newb in GitHub, so please forgive the following question.)  With that said, what does it mean to be assigned to Milestone 2.0 and seeing the (Preview) text Milestone section?  Willing to read up on the subject if pointed in the right direction. ;-)

With that question out of the way, what is the overall status of this?  Has there been any additional conversation and, if so, the outcome?  If not, what needs to be done to create additional movement?

**Comment by gafter** [comment]
Fixed in #26846 and #26948
